lms/djangoapps/bookmarks/models.py

```diff
@@ -36,23 +49,73 @@ def create(cls, bookmark_data):
         Raises:
             ItemNotFoundError: If no block exists for the usage_key.
         """
-        usage_key = bookmark_data.pop('usage_key')
+        data = copy.deepcopy(data)
+
+        usage_key = data.pop('usage_key')
         usage_key = usage_key.replace(course_key=modulestore().fill_in_run(usage_key.course_key))
 
         block = modulestore().get_item(usage_key)
 
-        bookmark_data['course_key'] = usage_key.course_key
-        bookmark_data['display_name'] = block.display_name
-        bookmark_data['path'] = cls.get_path(block)
-        user = bookmark_data.pop('user')
+        xblock_cache = XBlockCache.create({
+            'usage_key': usage_key,
+            'display_name': block.display_name,
+        })
 
-        bookmark, __ = cls.objects.get_or_create(usage_key=usage_key, user=user, defaults=bookmark_data)
+        data['course_key'] = usage_key.course_key
+        data['xblock_cache'] = xblock_cache
+
+        if xblock_cache.paths and len(xblock_cache.paths) == 1:
+            # If there is only one path to the block, use it.
+            data['path'] = xblock_cache.paths[0]
+        else:
+            # In case of multiple paths, in the future get_path_data()
+            # can be updated to look at which path the user has access to.
+            data['path'] = Bookmark.get_path_data(usage_key)
+
+        user = data.pop('user')
+
+        bookmark, __ = cls.objects.get_or_create(usage_key=usage_key, user=user, defaults=data)
         return bookmark
 
+    @property
+    def display_name(self):
+        """
+        Return the display_name from self.xblock_cache.
+
+        Returns:
+            String.
+        """
+        return self.xblock_cache.display_name
+
+    @property
+    def updated_path(self):
+        """
+        Return the latest path to this block after checking self.xblock_cache.
+
+        Returns:
+            List of dicts.
+        """
+        if self.modified < self.xblock_cache.modified:
+
+            if self.xblock_cache.paths and len(self.xblock_cache.paths) == 1:
```

Comment:
Nope, `xblock_cache.paths` is the list of all possible paths to the block. Its calculated by doing a DFS of the course tree without reference to any user.

On the other hand the path that `jump_to_id` and `path_to_location` construct is by doing `get_parent()`. This is also done without reference to any user and in case of DAGS, as far as I understand which path is returned depends on the modulestore.

So there can be two cases:
1. The DFS sets only one path in `xblock_cache.paths`. So this will also be the path `jump_to_id` and `path_to_location` return and we can just use it without any worry.
2. However, if the DFS sets more than one path in `xblock_cache.paths`, we don't know here which one the user will be taken to by `jump_to_id`, which is used by the UI. So we call `path_to_location` to get hold of this.

I'll add all this as comments.